File: src/lib/auction.ts
/**
 * 拍卖业务逻辑(状态推进 / 结束 / 生成胜出订单)
 *
 * 由于 Demo 没有定时任务,我们采用 lazy 推进策略:
 *   - 任意 GET / POST 涉及拍卖的请求,先调 advanceAuctionState() 把过期的拍卖关掉
 *   - 关闭时一次性确定 winner、生成订单、保证金抵扣 / 退还
 */

import { prisma } from './db';
import { Prisma, AuctionStatus, AuctionResult, DepositStatus, OrderSource, OrderStatus } from '@prisma/client';

/**
 * 把已经到期(>= endAt)且仍 live 的拍卖一次性收尾。
 * 同时:
 *   - 待开始 → 进行中
 *   - 已结束的 won 订单超过 24h 未付款 → defaulted(没收保证金)
 */
export async function advanceAuctionState(auctionId?: string) {
  const now = new Date();

  // 1) 待开始 → 进行中
  await prisma.auction.updateMany({
    where: {
      ...(auctionId ? { id: auctionId } : {}),
      status: AuctionStatus.scheduled,
      startAt: { lte: now },
    },
    data: { status: AuctionStatus.live },
  });

  // 2) 进行中 → 已结束(决出胜者 / 流拍)
  const expired = await prisma.auction.findMany({
    where: {
      ...(auctionId ? { id: auctionId } : {}),
      status: AuctionStatus.live,
      endAt: { lte: now },
    },
    select: { id: true },
  });
  for (const { id } of expired) {
    try {
      await finalizeAuction(id);
    } catch (e) {
      // eslint-disable-next-line no-console
      console.error(`[auction] finalize ${id} failed:`, e);
    }
  }

  // 3) 胜者 24h 未付款 → defaulted + 罚没保证金
  await enforceDefaultedAuctions(auctionId);
}

const DEFAULT_PAYMENT_DEADLINE_HOURS = 24;

/** 扫描所有「已胜出但订单仍 pending_payment」且超过 24h 的拍卖,标记违约 */
async function enforceDefaultedAuctions(auctionId?: string) {
  const cutoff = new Date(Date.now() - DEFAULT_PAYMENT_DEADLINE_HOURS * 3600_000);
  const candidates = await prisma.auction.findMany({
    where: {
      ...(auctionId ? { id: auctionId } : {}),
      status: AuctionStatus.finished,
      result: AuctionResult.won,
      actualEndAt: { lte: cutoff },
    },
    select: { id: true, winnerId: true, winningOrderId: true, title: true, sellerId: true, depositAmount: true },
  });
  if (candidates.length === 0) return;

  for (const a of candidates) {
    try {
      await defaultOneAuction(a);
    } catch (e) {
      // eslint-disable-next-line no-console
      console.error(`[auction] enforce default ${a.id} failed:`, e);
    }
  }
}

async function defaultOneAuction(a: {
  id: string;
  winnerId: string | null;
  winningOrderId: string | null;
  title: string;
  sellerId: string;
  depositAmount: number;
}) {
  if (!a.winningOrderId) return;
  const order = await prisma.order.findUnique({
    where: { id: a.winningOrderId },
    select: { id: true, status: true },
  });
  // 已支付 / 已取消 / 已退款 / 已发货 → 不需要违约处理
  if (!order || order.status !== 'pending_payment') return;

  await prisma.$transaction(async (tx) => {
    // 拍卖标记为违约
    await tx.auction.update({
      where: { id: a.id },
      data: { result: AuctionResult.defaulted },
    });
    // 订单作废(按 winningOrderId)
    await tx.order.update({
      where: { id: a.winningOrderId! },
      data: {
        status: 'cancelled',
        cancelledAt: new Date(),
        refundReason: '胜出者 24 小时内未付款',
      },
    });
    // 把对应参与者保证金标记为 forfeited(没收)
    if (a.winnerId) {
      await tx.auctionParticipant.updateMany({
        where: { auctionId: a.id, userId: a.winnerId, depositStatus: DepositStatus.applied },
        data: { depositStatus: DepositStatus.forfeited },
      });
    }
    // 通知胜者 + 卖家
    if (a.winnerId) {
      await tx.notification.create({
        data: {
          recipientId: a.winnerId,
          type: 'system',
          text: `❗ 你拍下的「${a.title}」未在 24 小时内完成付款,订单已取消,保证金 ¥${(a.depositAmount / 100).toFixed(2)} 已被没收`,
          link: `/auction/${a.id}`,
        },
      });
    }
    await tx.notification.create({
      data: {
        recipientId: a.sellerId,
        type: 'system',
        text: `⚠️ 你的拍卖品「${a.title}」胜出者未在 24 小时内付款,订单已取消`,
        link: `/auction/${a.id}`,
      },
    });
  });
}

/** 拍卖结束:决出胜出者、生成订单、处理保证金 */
async function finalizeAuction(auctionId: string) {
  await prisma.$transaction(async (tx) => {
    const auction = await tx.auction.findUnique({
      where: { id: auctionId },
      include: {
        bids: {
          orderBy: [{ amount: 'desc' }, { createdAt: 'asc' }],
          take: 1,
        },
      },
    });
    if (!auction || auction.status !== 'live') return;

    const now = new Date();
    const winningBid = auction.bids[0];
    const reached = !auction.reservePrice || (winningBid && winningBid.amount >= auction.reservePrice);

    if (!winningBid || !reached) {
      // 流拍
      await tx.auction.update({
        where: { id: auctionId },
        data: {
          status: AuctionStatus.finished,
          result: AuctionResult.no_bidder,
          actualEndAt: now,
        },
      });
      // 全部参与者保证金原路退回
      await refundAllDeposits(tx, auctionId);
      return;
    }

    const winnerId = winningBid.bidderId;

    // 生成胜出订单(尾款 = 中标价 - 已付保证金;但订单总金额 = 中标价,保证金记到 depositPaid)
    const orderNo = genOrderNo('AUC');
    const order = await tx.order.create({
      data: {
        orderNo,
        source: OrderSource.auction,
        auctionId: auction.id,
        productId: null,
        buyerId: winnerId,
        sellerId: auction.sellerId,
        quantity: 1,
        unitPrice: winningBid.amount,
        totalPrice: winningBid.amount,
        depositPaid: auction.depositAmount,    // 直接抵扣
        pointsBackTotal: 0,
        status: OrderStatus.pending_payment,
      },
    });

    await tx.auction.update({
      where: { id: auctionId },
      data: {
        status: AuctionStatus.finished,
        result: AuctionResult.won,
        winnerId,
        winningOrderId: order.id,
        actualEndAt: now,
        currentPrice: winningBid.amount,
      },
    });

    // 把胜出者的保证金标记为 applied(抵扣到这个订单)
    await tx.auctionParticipant.updateMany({
      where: { auctionId, userId: winnerId, depositStatus: DepositStatus.held },
      data: { depositStatus: DepositStatus.applied },
    });

    // 其他参与者的保证金原路退回
    await refundOtherParticipants(tx, auctionId, winnerId);

    // 通知胜出者
    await tx.notification.create({
      data: {
        recipientId: winnerId,
        type: 'system',
        text: `🎉 恭喜你以 ¥${(winningBid.amount / 100).toFixed(2)} 的价格拍下了「${auction.title}」,请在 24 小时内支付尾款 ¥${((winningBid.amount - auction.depositAmount) / 100).toFixed(2)}`,
        link: `/orders`,
      },
    });
    // 通知卖家
    await tx.notification.create({
      data: {
        recipientId: auction.sellerId,
        type: 'system',
        text: `🔨 你的拍卖品「${auction.title}」已成交,等待买家付款`,
        link: `/auction/${auction.id}`,
      },
    });
  });
}

/** 流拍时:全员退保证金 */
async function refundAllDeposits(tx: Prisma.TransactionClient, auctionId: string) {
  const ps = await tx.auctionParticipant.findMany({
    where: { auctionId, depositStatus: DepositStatus.held },
  });
  for (const p of ps) {
    await refundOneParticipant(tx, p.id, p.userId, p.depositAmount);
  }
}

/** 胜出后:除胜者外退保证金 */
async function refundOtherParticipants(
  tx: Prisma.TransactionClient,
  auctionId: string,
  winnerId: string
) {
  const ps = await tx.auctionParticipant.findMany({
    where: {
      auctionId,
      depositStatus: DepositStatus.held,
      userId: { not: winnerId },
    },
  });
  for (const p of ps) {
    await refundOneParticipant(tx, p.id, p.userId, p.depositAmount);
  }
}

async function refundOneParticipant(
  tx: Prisma.TransactionClient,
  participantId: string,
  userId: string,
  amount: number
) {
  await tx.auctionParticipant.update({
    where: { id: participantId },
    data: { depositStatus: DepositStatus.refunded },
  });
  // 保证金以积分形式退回(简化)。真实生产应通过原支付渠道退款。
  // 这里根据保证金支付时的渠道决定:Demo 中无论现金渠道,都按比例转积分
  // (1 元 = 100 积分),并发通知。
  const pts = Math.max(1, Math.floor(amount / 100) * 100);
  const u = await tx.user.update({
    where: { id: userId },
    data: { pointsBalance: { increment: pts } },
    select: { pointsBalance: true },
  });
  await tx.pointsLedger.create({
    data: {
      userId,
      type: 'admin',
      delta: pts,
      balance: u.pointsBalance,
      refType: 'auction_deposit_refund',
      refId: participantId,
      remark: `拍卖保证金退还(¥${(amount / 100).toFixed(2)})`,
    },
  });
  await tx.notification.create({
    data: {
      recipientId: userId,
      type: 'system',
      text: `💰 你的拍卖保证金 ¥${(amount / 100).toFixed(2)} 已退回(等值 ${pts} 积分)`,
    },
  });
}

export function genOrderNo(prefix = 'RY') {
  const now = new Date();
  const ymd =
    now.getFullYear().toString().slice(2) +
    String(now.getMonth() + 1).padStart(2, '0') +
    String(now.getDate()).padStart(2, '0');
  return (
    prefix +
    ymd +
    Date.now().toString(36).slice(-5).toUpperCase() +
    Math.floor(Math.random() * 1000).toString().padStart(3, '0')
  );
}

File: server-go/internal/auction/auction.go
// Package auction 对齐 Next.js src/lib/auction.ts 的拍卖状态机。
// 策略:lazy 推进 — 任何读写拍卖的请求先调 AdvanceState。
package auction

import (
	"fmt"
	"log"
	"time"

	"plants-community-server/internal/models"

	"gorm.io/gorm"
)

const DefaultPaymentDeadlineHours = 24

// AdvanceState 把所有过期的 live 拍卖收尾,同时把 scheduled → live,
// 并对「胜出但 24h 未付款」的订单执行违约处理。
// 如 auctionID != "",则只处理该场。
func AdvanceState(db *gorm.DB, auctionID string) {
	now := time.Now()

	// 1) scheduled → live
	q := db.Model(&models.Auction{}).
		Where("status = ? AND startAt <= ?", "scheduled", now)
	if auctionID != "" {
		q = q.Where("id = ?", auctionID)
	}
	q.Update("status", "live")

	// 2) live → finished(决出胜者 / 流拍)
	type row struct{ ID string }
	var expired []row
	qe := db.Model(&models.Auction{}).
		Where("status = ? AND endAt <= ?", "live", now)
	if auctionID != "" {
		qe = qe.Where("id = ?", auctionID)
	}
	qe.Select("id").Scan(&expired)
	for _, r := range expired {
		if err := finalize(db, r.ID); err != nil {
			log.Printf("[auction] finalize %s failed: %v", r.ID, err)
		}
	}

	// 3) 违约扫描
	enforceDefaulted(db, auctionID)
}

// finalize 拍卖结束处理
func finalize(db *gorm.DB, auctionID string) error {
	return db.Transaction(func(tx *gorm.DB) error {
		var a models.Auction
		if err := tx.Where("id = ?", auctionID).First(&a).Error; err != nil {
			return err
		}
		if a.Status != "live" {
			return nil
		}

		// 查最高出价
		var topBid models.Bid
		hasBid := tx.Where("auctionId = ?", auctionID).
			Order("amount DESC, createdAt ASC").First(&topBid).Error == nil

		// 保留价校验
		reached := a.ReservePrice == nil || (hasBid && topBid.Amount >= *a.ReservePrice)
		now := time.Now()

		if !hasBid || !reached {
			// 流拍
			tx.Model(&a).Updates(map[string]any{
				"status": "finished", "result": "no_bidder", "actualEndAt": &now,
			})
			refundAllDeposits(tx, auctionID)
			return nil
		}

		// 成交
		winnerID := topBid.BidderID
		orderID := genCuidAuction()
		orderNo := genAuctionOrderNo()
		order := models.Order{
			ID: orderID, OrderNo: orderNo,
			Source: "auction", AuctionID: &a.ID, ProductID: nil,
			BuyerID: winnerID, SellerID: &a.SellerID,
			Quantity: 1, UnitPrice: topBid.Amount,
			TotalPrice: topBid.Amount, DepositPaid: a.DepositAmount,
			PointsBackTotal: 0, Status: "pending_payment",
			CreatedAt: now, UpdatedAt: now,
		}
		if err := tx.Create(&order).Error; err != nil {
			return err
		}

		tx.Model(&a).Updates(map[string]any{
			"status": "finished", "result": "won",
			"winnerId":       winnerID,
			"winningOrderId": orderID,
			"actualEndAt":    &now,
			"currentPrice":   topBid.Amount,
		})

		// 胜者保证金标记为 applied
		tx.Model(&models.AuctionParticipant{}).
			Where("auctionId = ? AND userId = ? AND depositStatus = ?",
				auctionID, winnerID, "held").
			Update("depositStatus", "applied")

		// 其他人退保证金
		refundOthers(tx, auctionID, winnerID)

		// 通知
		balance := topBid.Amount - a.DepositAmount
		tx.Create(&models.Notification{
			ID: genCuidAuction(), RecipientID: winnerID, Type: "system",
			Text: fmt.Sprintf("🎉 恭喜你以 ¥%.2f 的价格拍下了「%s」,请在 24 小时内支付尾款 ¥%.2f",
				float64(topBid.Amount)/100, a.Title, float64(balance)/100),
			Link:      ptr("/orders"),
			CreatedAt: now,
		})
		tx.Create(&models.Notification{
			ID: genCuidAuction(), RecipientID: a.SellerID, Type: "system",
			Text:      fmt.Sprintf("🔨 你的拍卖品「%s」已成交,等待买家付款", a.Title),
			Link:      ptr("/auction/" + a.ID),
			CreatedAt: now,
		})
		return nil
	})
}

// 违约扫描
func enforceDefaulted(db *gorm.DB, auctionID string) {
	cutoff := time.Now().Add(-time.Duration(DefaultPaymentDeadlineHours) * time.Hour)
	q := db.Model(&models.Auction{}).
		Where("status = ? AND result = ? AND actualEndAt <= ?",
			"finished", "won", cutoff)
	if auctionID != "" {
		q = q.Where("id = ?", auctionID)
	}
	var rows []models.Auction
	q.Find(&rows)
	for _, a := range rows {
		if a.WinningOrderID == nil {
			continue
		}
		var o models.Order
		if db.Where("id = ?", *a.WinningOrderID).First(&o).Error != nil {
			continue
		}
		if o.Status != "pending_payment" {
			continue
		}
		defaultOne(db, &a, &o)
	}
}

func defaultOne(db *gorm.DB, a *models.Auction, o *models.Order) {
	db.Transaction(func(tx *gorm.DB) error {
		// 拍卖 defaulted
		tx.Model(a).Update("result", "defaulted")
		// 订单取消
		now := time.Now()
		tx.Model(o).Updates(map[string]any{
			"status": "cancelled", "cancelledAt": &now,
			"refundReason": "胜出者 24 小时内未付款",
		})
		// 保证金没收
		if a.WinnerID != nil {
			tx.Model(&models.AuctionParticipant{}).
				Where("auctionId = ? AND userId = ? AND depositStatus = ?",
					a.ID, *a.WinnerID, "applied").
				Update("depositStatus", "forfeited")

			tx.Create(&models.Notification{
				ID: genCuidAuction(), RecipientID: *a.WinnerID, Type: "system",
				Text: fmt.Sprintf("❗ 你拍下的「%s」未在 24 小时内完成付款,订单已取消,保证金 ¥%.2f 已被没收",
					a.Title, float64(a.DepositAmount)/100),
				Link: ptr("/auction/" + a.ID), CreatedAt: now,
			})
		}
		tx.Create(&models.Notification{
			ID: genCuidAuction(), RecipientID: a.SellerID, Type: "system",
			Text:      fmt.Sprintf("⚠️ 你的拍卖品「%s」胜出者未在 24 小时内付款,订单已取消", a.Title),
			Link:      ptr("/auction/" + a.ID),
			CreatedAt: now,
		})
		return nil
	})
}

// refundAllDeposits 流拍时全退
func refundAllDeposits(tx *gorm.DB, auctionID string) {
	var ps []models.AuctionParticipant
	tx.Where("auctionId = ? AND depositStatus = ?", auctionID, "held").Find(&ps)
	for _, p := range ps {
		refundOne(tx, &p)
	}
}

// refundOthers 成交时除胜者外全退
func refundOthers(tx *gorm.DB, auctionID, winnerID string) {
	var ps []models.AuctionParticipant
	tx.Where("auctionId = ? AND depositStatus = ? AND userId <> ?",
		auctionID, "held", winnerID).Find(&ps)
	for _, p := range ps {
		refundOne(tx, &p)
	}
}

// refundOne 退保证金:标记 refunded,等值钻石回到账户,发通知
func refundOne(tx *gorm.DB, p *models.AuctionParticipant) {
	tx.Model(p).Update("depositStatus", "refunded")
	// 1元 = 100 钻石(与 TS 版一致)
	pts := p.DepositAmount / 100 * 100
	if pts < 1 {
		pts = 1
	}
	tx.Model(&models.User{}).Where("id = ?", p.UserID).
		UpdateColumn("pointsBalance", gorm.Expr("pointsBalance + ?", pts))
	var u models.User
	tx.Where("id = ?", p.UserID).First(&u)
	tx.Create(&models.PointsLedger{
		ID: genCuidAuction(), UserID: p.UserID, Type: "admin",
		Delta: pts, Balance: u.PointsBalance,
		RefType: ptr("auction_deposit_refund"), RefID: ptr(p.ID),
		Remark:    ptr(fmt.Sprintf("拍卖保证金退还(¥%.2f)", float64(p.DepositAmount)/100)),
		CreatedAt: time.Now(),
	})
	tx.Create(&models.Notification{
		ID: genCuidAuction(), RecipientID: p.UserID, Type: "system",
		Text: fmt.Sprintf("💰 你的拍卖保证金 ¥%.2f 已退回(等值 %d 钻石)",
			float64(p.DepositAmount)/100, pts),
		CreatedAt: time.Now(),
	})
}

// ===== utils =====
func ptr(s string) *string { return &s }

// 为了避免循环导入,这里复写简版 cuid/orderNo
func genCuidAuction() string {
	return fmt.Sprintf("c%d%d", time.Now().UnixNano(), time.Now().Nanosecond()%10000)
}

func genAuctionOrderNo() string {
	now := time.Now()
	ymd := fmt.Sprintf("%02d%02d%02d", now.Year()%100, int(now.Month()), now.Day())
	t36 := fmt.Sprintf("%X", now.UnixNano())
	if len(t36) > 5 {
		t36 = t36[len(t36)-5:]
	}
	return "AUC" + ymd + t36 + fmt.Sprintf("%03d", now.Nanosecond()%1000)
}
